### DISCUSSION TITLE
Support for NXP's Mifare DESFire

### DISCUSSION THREAD

**Discussion by theo-basty** [OpenSC/OpenSC#2797] [Ideas]
I've started to dig in the deephole that is smartcard based authentication on Windows.

The idea behind that is to be able to authenticate users in harsh factory environment while they are accessing softwares from within the production areas. 

The users there often have protective gear which makes it hard to enter their credentials, so i was thinking of using smart card login to make it easier for them. The idee would to use Windows Smartcard Logon so that we can manage this card through the domain and have a better software compatibility from the get go.

However, because some area are falling under ATEX regulations (explosive atmosphere) we cannot contact based token because electrical connection in these environments is prohibited.

While searching for contactless based smartcards I often ended up on Mifare DESFire EV2 or EV3 tokens/cards, which seems to be quite cheap, could also be used for existing physical access control and seems to support encryption algorithms. 

However I can't seem to fin either minidrivers or CSPs provided by NXP or anyone else to use these token/cards with Windows Smartcard Logon.

How much would it be to add support for these token in OpenSC ? If it is possible to add this support, it would be a huge help to bring cheap and easily available NFC smartcard logon for Windows.

The official chip product page :
https://www.nxp.com/products/rfid-nfc/mifare-hf/mifare-desfire/mifare-desfire-ev3-high-security-ic-for-contactless-smart-city-services:MF3DHx3

**Comment by popovec** [comment]
Mifare DESFire EV2 and EV3 do not support asymmetric encryption. They don't have support for RSA or elliptic encryption or any of the post quantum asymmetric ciphers (which could be used for digital signature or decryption or key exchange). 

According to the documentation, neither MSE operation (ins 0x22) nor PSO operation (ins 0x2a) is available. Without these operations, the listed Mifare card is practically outside the scope of OpenSC project.

It is possible to use these cards for identity verification, but I personally do not know a reasonable (and safe) methodology that allows to solve the revocation of cards, the life cycle of cards without asymmetric encryption/without PKI.

**Reply by theo-basty** [reply]
Oh ! I see... I didn't understand that asymmetric encryption was needed for the full scope of OpenSC features.

Thank you for the answer ! I will need to check for other options then.

**Comment by frankmorgner** [comment]
Use https://github.com/nfc-tools/libfreefare to access your DESFire card. Also, you might want to check out https://www.mysmartlogon.com/products/nfccsp.html with support for 14443 tags.

**Comment by AlexDev404** [comment]
> I've started to dig in the deephole that is smartcard based authentication on Windows.
> 
> The idea behind that is to be able to authenticate users in harsh factory environment while they are accessing softwares from within the production areas.
> 
> The users there often have protective gear which makes it hard to enter their credentials, so i was thinking of using smart card login to make it easier for them. The idee would to use Windows Smartcard Logon so that we can manage this card through the domain and have a better software compatibility from the get go.
> 
> However, because some area are falling under ATEX regulations (explosive atmosphere) we cannot contact based token because electrical connection in these environments is prohibited.
> 
> While searching for contactless based smartcards I often ended up on Mifare DESFire EV2 or EV3 tokens/cards, which seems to be quite cheap, could also be used for existing physical access control and seems to support encryption algorithms.
> 
> However I can't seem to fin either minidrivers or CSPs provided by NXP or anyone else to use these token/cards with Windows Smartcard Logon.
> 
> How much would it be to add support for these token in OpenSC ? If it is possible to add this support, it would be a huge help to bring cheap and easily available NFC smartcard logon for Windows.
> 
> The official chip product page : https://www.nxp.com/products/rfid-nfc/mifare-hf/mifare-desfire/mifare-desfire-ev3-high-security-ic-for-contactless-smart-city-services:MF3DHx3

You can try https://github.com/ObsidianPay/libnxpsc if you're interested. It works like this:

```C++
#include <winscard.h>
#include <stdio.h>

extern "C" {
#include <nxpsc/nxpsc.h>
}

static int pcsc_transceive(void* vctx, const uint8_t* tx, size_t txlen,
    uint8_t* rx, size_t rxcap, size_t* rxlen);

struct PcscCtx {
    SCARDHANDLE hCard;
    DWORD proto;
};


int main()
{

    printf("libnxpsc version: %s\n", nxpsc_version_string());
    // Library selftest
    if (nxpsc_selftest(true) != NXPSC_OK) {
        printf("Library selftest failed\n");
        return 1;
    }
    PcscCtx ctx;
    SCARDCONTEXT hContext;
    LONG rv = SCardEstablishContext(SCARD_SCOPE_USER, NULL, NULL, &hContext);
    if (rv != SCARD_S_SUCCESS) {
        printf("Failed to establish context: %ld\n", rv);
        return 1;
    }

    // List available readers
    LPTSTR mszReaders = NULL;
    DWORD dwReaders = SCARD_AUTOALLOCATE;
    rv = SCardListReaders(hContext, NULL, (LPTSTR)&mszReaders, &dwReaders);
    if (rv != SCARD_S_SUCCESS) {
        printf("Failed to list readers: %ld\n", rv);
        SCardReleaseContext(hContext);
        return 1;
    }

    // Connect to the first reader
    rv = SCardConnect(hContext, mszReaders, SCARD_SHARE_SHARED, SCARD_PROTOCOL_T0 | SCARD_PROTOCOL_T1, &ctx.hCard, &ctx.proto);
    if (rv != SCARD_S_SUCCESS) {
        printf("no card or bad reader\n");
        SCardFreeMemory(hContext, mszReaders);
        SCardReleaseContext(hContext);
        return 1;
    }

    nxpsc_transport_t transport = {};
    transport.ctx = &ctx;
    transport.transceive = pcsc_transceive;

    nxpsc_card_t* card;
    int rc = nxpsc_open(&transport, &card);
    if (rc != NXPSC_OK && nxpsc_selftest(true)) {
        printf("Failed to open card: %d\n", rc);
        SCardDisconnect(ctx.hCard, SCARD_LEAVE_CARD);
        SCardReleaseContext(hContext);
        return 1;
    }
    printf("\nthe card is connected\n\n");

    // Identify the card
    nxpsc_cardtype_t type = NXP_UNKNOWN;
    nxpsc_identify(card, &type);

    printf("Card type: %s\n", nxpsc_cardtype_str(type));

    nxpsc_version_t v{};
    rc = nxpsc_get_version(card, &v);
    if (rc == NXPSC_OK) {
        printf("sw version: %u.%u\n", v.sw_major, v.sw_minor);
        printf("hw type: %02X\n", v.hw_type);
        printf("hw storage size: %u bytes\n", v.hw_storage);
        printf("hw protocol: %02X\n", v.hw_protocol);
    }

    // Bytes free
    uint32_t bytes = 0;
    rc = nxpsc_get_free_memory(card, &bytes);
    if (rc == NXPSC_OK)
        printf("free: %u bytes\n", bytes);
    return 0;
}

static int pcsc_transceive(void* vctx, const uint8_t* tx, size_t txlen,
    uint8_t* rx, size_t rxcap, size_t* rxlen)
{
    PcscCtx* ctx = static_cast<PcscCtx*>(vctx);
    DWORD rlen = static_cast<DWORD>(rxcap);

    // Debug output of the sent data
    printf(">> ");
    for (size_t i = 0; i < txlen; i++) printf("%02X", tx[i]);
    printf("\n");

    // Transmit the APDU to the card
    LONG rc = SCardTransmit(ctx->hCard, SCARD_PCI_T1,
        tx, static_cast<DWORD>(txlen),
        nullptr, rx, &rlen);
    if (rc != SCARD_S_SUCCESS)
        return NXPSC_E_TRANSPORT;   // card removed / no answer

    // Debug output of the received data
    printf("<< ");
    for (DWORD i = 0; i < rlen; i++) printf("%02X", rx[i]);
    printf("\n\n");

    *rxlen = rlen;
    return NXPSC_OK;                // card errors are decoded by the library
}
```

**Comment by frankmorgner** [comment]
My [response above](https://github.com/OpenSC/OpenSC/discussions/2797#discussioncomment-7619118) still holds. You can do DESFire with a Windows Credential Provider as done in https://www.mysmartlogon.com/products/nfccsp.html, but you cannot use OpenSC for this.

**Comment by AlexDev404** [comment]
> My [response above](https://github.com/OpenSC/OpenSC/discussions/2797#discussioncomment-7619118) still holds. You can do DESFire with a Windows Credential Provider as done in https://www.mysmartlogon.com/products/nfccsp.html, but you cannot use OpenSC for this.

I was speaking to him on something completely different, but still related, on how to interact with the cards directly so that he doesn't rely on OpenSC at all.

**Comment by dengert** [comment]
Your other option if you are using Windows AD is to use contactless (NFC)  smart cards and contactless readers, they would still have to type in a pin.  But the contactless reader is usually in a USB port, which may not meet your requirements. 

Yubico tokens also support contactless. They can be provisioned via your Window AD managers using contact, then used in your contactless environment.  You can also set up some smart cards to not require a PIN for authentication, which is authenticating the card - not the user.  


Some cards like can do authentication of the user over contactless  such as   
NIST PIV cards with optional NIST Secure Messaging can also be used of contactless interface.

**Comment by dengert** [comment]
https://www.mysmartlogon.com/products/nfccsp.html might work for you. My response was to use some smart cards with their Windows minidriver or OpenSC's minidriver to get full AD security access. But in your environment that may not be needed.

**Reply by AlexDev404** [reply]
> https://www.mysmartlogon.com/products/nfccsp.html might work for you. My response was to use some smart cards with their Windows minidriver or OpenSC's minidriver to get full AD security access. But in your environment that may not be needed.     

Hey not to be rude, but I'm not sure what you guys are referring to in this link. You all are passing around this but it just leads to a 404.

Also with this software you need to pay and it's a completely proprietary software.

**Comment by dengert** [comment]
the 404 page I saw would still let you look at other pages including some
with NFC

**Reply by AlexDev404** [reply]
So if you're aware that it's a 404 page then why post it at all? Better to not post anything or instead, find and send the correct link
